Ensure devices are preserved when forwarding between futures

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57443 Make RRefContext get devices from RPC agent when creating OwnerRRef
* #57442 Replace DeviceIndexes with Devices in RRefs
* #56918 Make DataPtr extraction in CUDAFuture faster for Python values
* **#57432 Ensure devices are preserved when forwarding between futures**
* #57433 Avoid re-extracting DataPtrs when forwarding values between Futures

In a bunch of places we were creating a future and then "forwarding" the value of another future to it once that other future completed. (This was in order to convert the type of the value, or to "merge" multiple futures into one). However when doing so we often created a child future with an empty set of devices, which meant it didn't support CUDA, and thus would cause a silent synchronization/correctness bug if the parent future did actually contain CUDA tensors.

One way this could have been caught earlier would have been to have Future always extract the DataPtrs, even in CPU-only mode, in order to ensure they always reside on the expected set of devices. Unfortunately this might have some averse perf effects thus should be done carefully.

Differential Revision: [D28143045](https://our.internmc.facebook.com/intern/diff/D28143045/)

lw committed:
Ensure devices are preserved when forwarding between futures

In a bunch of places we were creating a future and then "forwarding" the value of another future to it once that other future completed. (This was in order to convert the type of the value, or to "merge" multiple futures into one). However when doing so we often created a child future with an empty set of devices, which meant it didn't support CUDA, and thus would cause a silent synchronization/correctness bug if the parent future did actually contain CUDA tensors.

One way this could have been caught earlier would have been to have Future always extract the DataPtrs, even in CPU-only mode, in order to ensure they always reside on the expected set of devices. Unfortunately this might have some averse perf effects thus should be done carefully.

Differential Revision: [D28143045](https://our.internmc.facebook.com/intern/diff/D28143045/)

[ghstack-poisoned]

lw committed:
Update on "Ensure devices are preserved when forwarding between futures"


In a bunch of places we were creating a future and then "forwarding" the value of another future to it once that other future completed. (This was in order to convert the type of the value, or to "merge" multiple futures into one). However when doing so we often created a child future with an empty set of devices, which meant it didn't support CUDA, and thus would cause a silent synchronization/correctness bug if the parent future did actually contain CUDA tensors.

One way this could have been caught earlier would have been to have Future always extract the DataPtrs, even in CPU-only mode, in order to ensure they always reside on the expected set of devices. Unfortunately this might have some averse perf effects thus should be done carefully.

Differential Revision: [D28143045](https://our.internmc.facebook.com/intern/diff/D28143045/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -1009,21 +1011,23 @@ TORCH_API intrusive_ptr<ivalue::Future> collectAny(
       return srcs.get(i);
     }
     TORCH_CHECK(i == 0 || (*typePtr == *srcs.get(i)->elementType()));
+    TORCH_CHECK(i == 0 || (devices == srcs.get(i)->devices()));
```
This is prior to this PR: `TORCH_CHECK` are usually user-facing checks with clear error messages. If this is internal only, maybe use `TORCH_INTERNAL_ASSERT` instead?

mrshenli replied:
This requires all Futures' value using the same set of devices? I guess this might be fine as we haven't exposed this to Python land. I am not sure if there are existing C++ use-cases of this API. cc @pritamdamania87

lw replied:
> TORCH_CHECK are usually user-facing checks

I think this might be user-facing? I mean, the error happens if the user passes in bad values, which could totally happen. I don't know if the end user could call this directly, or if it's private, but within this function this is just an argument validation error. I can put a clearer error message.

> This requires all Futures' value using the same set of devices?

Alternatively we could make this function compute the union of the set of devices of all of its arguments and use that. I went for this other way because it was simpler and I expected it to be enough in practice (e.g., all futures obtained from RPC have the same set of devices). But it'll be very easy to relax this restriction if we'll need to.

pritamdamania87 replied:
> I am not sure if there are existing C++ use-cases of this API. cc @pritamdamania87

Not aware of any usage of this API yet

mrshenli commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -1009,21 +1011,23 @@ TORCH_API intrusive_ptr<ivalue::Future> collectAny(
   }
   struct Ctx {
-    explicit Ctx(List<intrusive_ptr<ivalue::Future>> srcs, TypePtr typePtr)
+    explicit Ctx(List<intrusive_ptr<ivalue::Future>> srcs, TypePtr typePtr, std::vector<c10::Device> devices)
```
let's break long line

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -58,7 +58,8 @@ std::shared_ptr<JitFuture> RpcAgent::sendWithRetries(
 
-  auto originalFuture = std::make_shared<JitFuture>(at::AnyClassType::get());
+  auto originalFuture =
+      std::make_shared<JitFuture>(at::AnyClassType::get(), getDevices());
```
I assume we need `getDevices()` here so that `TensorPipeAgent` can override this API and enforce its own devices.

lw replied:
Yes indeed here it made sense to me to create a future that had the same set of devices as what the agent supports, hence we needed a generic way to retrieve that set of devices. It also so happened that, for later in the stack, I also needed a way to retrieve the set of devices from the agent.

Another way we could have done this, by reordering this code, is to do `originalFuture = jitFuture->createInstance(at::AnyClassType::get());` I could go that way if you prefer.

lw committed:
Update on "Ensure devices are preserved when forwarding between futures"


In a bunch of places we were creating a future and then "forwarding" the value of another future to it once that other future completed. (This was in order to convert the type of the value, or to "merge" multiple futures into one). However when doing so we often created a child future with an empty set of devices, which meant it didn't support CUDA, and thus would cause a silent synchronization/correctness bug if the parent future did actually contain CUDA tensors.

One way this could have been caught earlier would have been to have Future always extract the DataPtrs, even in CPU-only mode, in order to ensure they always reside on the expected set of devices. Unfortunately this might have some averse perf effects thus should be done carefully.

Differential Revision: [D28143045](https://our.internmc.facebook.com/intern/diff/D28143045/)

[ghstack-poisoned]
